---
title: "01_MLN"
author: "Ananya Gupta"
date: "2025-11-22"
output: html_document
---


```{r env_loading, results='FALSE',}
# Load packages, data and functions
library(Seurat)
library(tidyverse)
library(SeuratWrappers)
library(magicBatch)
library(BPCells)
library(batchelor)
library(patchwork)
library(plotly)
library(gprofiler2)
library(hdf5r)
library(kableExtra)
library(ggplot2)
library(gridExtra)
library(DT)
library(Cairo)
library(clustree)
library(Matrix)
library(scales)
library(data.table)
library(dplyr)
library(tidyverse)
library(parallelly)
library(glue)
library(ggrepel)
library(gprofiler2)
library(multtest)
library(metap)
library(future)
plan(multicore, workers = 8)


#for parallelization
options(future.globals.maxSize = 8000 * 1024^2) 
#options(future.globals.maxSize = 1e9) was abt 1GB


# Set the random number seed
set.seed(1234)
```


```{r dir setup, include=FALSE}
base_dir <- "MLN_outputs"

# Define output directories
rds_dir  <- file.path(base_dir, "MLN_rds")
fig_dir  <- file.path(base_dir, "MLN_figs")

# Create folders if they don't exist
dir.create(base_dir, showWarnings = FALSE)
dir.create(rds_dir, showWarnings = FALSE)
dir.create(fig_dir, showWarnings = FALSE)
```
 saving a fig ggsave(
  filename = file.path(fig_dir, "cluster_umap.png"),
  plot = p,
  width = 6,
  height = 5,
  dpi = 300
)


saveRDS(
  object = seu,
  file = file.path(rds_dir, "seu_pp_all_reps.rds")
)


p <- ggplot(mtcars, aes(mpg, wt)) + geom_point()

ggsave(
  filename = file.path(fig_dir, "myplot.pdf"),
  plot = p,
  device = CairoPDF,
  width = 6,
  height = 4,
  units = "in"
)

1 of ko,oe,wt,crectrl

```{r}
samples_df <- read.csv("samples_MLN.csv",header=TRUE, stringsAsFactors = FALSE) #gsub works on chr

samples_df$SampleName <- gsub("_", "-", samples_df$SampleName)

#fix path for windows
samples_df$Path <- trimws(samples_df$Path) #white space
samples_df$Path <-  normalizePath(samples_df$Path, winslash="/")#,mustWork = FALSE)


#setnames(values,namestoset)
datadirs <- setNames(samples_df[,4], samples_df[,1])

print(samples_df)
print(datadirs)

```

```{r}
seurat_list <- list()

for(i in 1:nrow(samples_df)) {
  sample_name <- samples_df$SampleName[i] #file names
  sample_group <- samples_df$SampleGroup[i] #genotype
  data_path <- samples_df$Path[i]
  
  cat("Processing:", sample_name, "\n")
  
  data <- Read10X(data.dir = data_path)
  
  #HTO data is multi output so it will be a list
  if(is.list(data)) {
    # For Cell Ranger multi output, extract only Gene Expression
    gene_exp <- data$`Gene Expression`
    cat("  - Multi output detected, using Gene Expression only\n")
  } else {
    # Regular output
    gene_exp <- data
    cat("  - Regular output\n")
  }
  
  # Create Seurat object
  seurat_obj <- CreateSeuratObject(
    counts = gene_exp,
    project = sample_name,
    min.cells = 3,
    min.features = 100
  )
  
  #metadata
  #seurat_obj$SampleName <- sample_name
  seurat_obj$genotype <- sample_group
  seurat_obj$site <- samples_df$Tissue[i]
  seurat_obj$condition <- samples_df$Condition[i]
  
  #qc
  seurat_obj$percent.mt <- PercentageFeatureSet(seurat_obj, pattern = "^mt-")

  
  # Store in list
  seurat_list[[sample_name]] <- seurat_obj
}

```

```{r}
# Merge all objects into one Seurat object
seu <- merge(
  x = seurat_list[[1]],
  y = seurat_list[2:length(seurat_list)],
  add.cell.ids = names(seurat_list),
  project = "mln_coup",merge.data = TRUE
)
```

```{r}
#library(Cairo)
save_pub <- function(plot,
                     filename,
                     width = 7,
                     height = 7,
                     dpi = 600) {

  

  # ---- SAVE PDF ----
  ggsave(
    filename = file.path(fig_dir, paste0(filename, ".pdf")),
    plot = plot,
    device = CairoPDF,
    width = width,
    height = height,
    units = "in"
  )

  # ---- SAVE PNG ----
  CairoPNG(
    filename = file.path(fig_dir, paste0(filename, ".png")),
    width  = width * dpi,   # convert inches → pixels
    height = height * dpi,
    res    = dpi
  )

  print(plot)    # this renders patchwork/Seurat correctly
  dev.off()

  return(plot)   # so you still see it in the RMarkdown output
}


```



```{r, fig.height=6,fig.width=7}
#Features above 6000 might be doublets ngl no but hevs have v high transcription

p <- FeatureScatter(
  seu,
  feature1 = "nCount_RNA",
  feature2 = "nFeature_RNA",
  group.by = "genotype"
) +
  plot_annotation(
    title = "MLN: UMIs vs Genes",
    theme = theme(
      plot.title = element_text(
        hjust = 0.5,
        size = 18,
        face = "bold"
      )
    )
  )

p
  #+ ggtitle("QC: UMIs vs Genes")
```
```{r}
save_pub(plot=p,  filename= "MLN_UMIs_vs_Genes", width=7, height=6)
```


```{r}
ncol(seu)
```

16467 cells

```{r}
table(seu$orig.ident)
```

COUPOE-MLN             COUPWT-MLN MLN-COUPKO-HT3-cre-neg MLN-COUPKO-HT5-cre-pos 
                  9422                   6809                    105                    131 
                  
```{r}
rm(seurat_obj)
rm(seurat_list)
```

```{r,fig.width=12}
df <- seu@meta.data
df$sum <- seu$nCount_RNA
df$detected <- seu$nFeature_RNA


p <- ggplot(df, aes(x = sum, y = detected)) +
  geom_point(aes(colour = percent.mt > 10), alpha = 0.7, size = 1) +
  facet_wrap(vars(genotype)) +
  scale_color_manual(
    values = c("FALSE" = "coral", "TRUE" = "turquoise3"),
    labels = c("FALSE" = "≤10%", "TRUE" = ">10%")
  ) +
  labs(
    title = "QC: Detected genes vs total counts per MLN Sample",
    x = "Total counts per cell",
    y = "Detected genes per cell",
    colour = "Mitochondrial %"
  ) +
  theme_bw() +
  theme(
    plot.title = element_text(hjust = 0.5, face = "bold"),
    strip.text = element_text(face = "bold")
  )
p
save_pub(plot=p,  filename= "genes_tot_count_mln", width=12, height=7)
```

```{r,fig.width=10, fig.height=4}
p <- VlnPlot(seu, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), group.by = "genotype",
              pt.size = 0,
              ncol = 3) +
  plot_annotation(
    title = "MLN: Before QC",
    theme = theme(
      plot.title = element_text(
        hjust = 0.5,
        size = 18,
        face = "bold"
      )
    )
  )

p
```

```{r}
save_pub(plot=p,  filename= "beforeqc_mln", width=10, height=4)
```


```{r}
#cleaning
seu <- subset(seu, subset = nFeature_RNA > 100 & percent.mt < 10 & nCount_RNA < 30000)
```

```{r}
rm(seurat_list)
```


```{r,fig.width=10, fig.height=4}
p <- VlnPlot(seu, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), group.by = "genotype",
              pt.size = 0,
              ncol = 3)  

p + plot_annotation(
    title = "MLN data After QC",
    theme = theme(
      plot.title = element_text(hjust = 0.5, size = 18, face = "bold")
    )
)

# ---- SAVE AS PDF ----
ggsave(
  filename = file.path(fig_dir, "MLN_data_After_QC.pdf"),
  plot = p,
  device = CairoPDF,
  width = 10,
  height = 4,
  units = "in"
)

# SAFEST WAY TO SAVE PNG ON WINDOWS
CairoPNG(
  filename = file.path(fig_dir, "MLN_data_After_QC.png"),
  width = 10 * 600,     # width in pixels (10 inches × 600 dpi)
  height = 4 * 600,     # height in pixels
  res = 600
)
print(p)        # <<--- VERY IMPORTANT for patchwork/ggplots
dev.off()
```
```{r}
save_pub(plot=p,  filename= "MLN_data_After_QC", width=10, height=4)
```

```{r}
ncol(seu) #22643
table(seu$orig.ident)
```
 COUPOE-MLN             COUPWT-MLN MLN-COUPKO-HT3-cre-neg MLN-COUPKO-HT5-cre-pos 
                  9116                   6747                    103                    128 

```{r}
seu <- NormalizeData(seu)
seu <- FindVariableFeatures(seu, nfeatures = 3000,verbose=F)
seu <- ScaleData(seu, verbose=F)
seu <- RunPCA(seu)
```

```{r}
saveRDS(
  object = seu,
  file = file.path(rds_dir, "seu_mln_gen_pca.rds"))
  
  
# Load it back
#seu <- readRDS(file.path(rds_dir, "seu_pp_gen_pca.rds"))
```

```{r}

# Load it back
seu <- readRDS(file.path(rds_dir, "seu_mln_gen_pca.rds"))
```


```{r}
seu <- IntegrateLayers(seu, method = FastMNNIntegration, batch=seu$orig.ident,features=VariableFeatures(seu), new.reduction = "integrated.mnn",verbose = TRUE)
```


```{r}
seu <- JoinLayers(seu)
```

```{r}
saveRDS(
  object = seu,
  file = file.path(rds_dir, "mln_int_join_layers.rds")) # just in case
#   
#   
# # Load it back
# #seu <- readRDS(file.path(rds_dir, "pp_int_join_layers.rds"))
```

First Lets do a diagnostic resolution study

```{r}
#seu <- readRDS(file.path(rds_dir, "mln_int_join_layers.rds"))
```

```{r}
ElbowPlot(seu, ndims= 50)
```

```{r}
# https://github.com/FrancisCrickInstitute/snRNAseq_workflow/blob/dabf3d87226ef9e9d51b12689c05a515cd1f0abc/templates/integrating_mnn.rmd
DefaultAssay(seu) <- "mnn.reconstructed"
```


```{r}
s.genes = gorth(cc.genes.updated.2019$s.genes, source_organism = "hsapiens", target_organism = "mmusculus")$ortholog_name
g2m.genes = gorth(cc.genes.updated.2019$g2m.genes, source_organism = "hsapiens", target_organism = "mmusculus")$ortholog_name

seu <- CellCycleScoring(object=seu, s.features=s.genes, g2m.features=g2m.genes, set.ident=F) #set in the meta.data col phase

```


```{r}
seu <- ScaleData(seu, vars.to.regress = c("S.Score", "G2M.Score"))
```

check karo does cell cell effect get removed? THEY DOOOO


```{r}
# when running a PCA on cell cycle genes, actively proliferating cells remain distinct from G1
# cells however, within actively proliferating cells, G2M and S phase cells group together
seu_obj <- RunPCA(seu, features = c(s.genes, g2m.genes))
DimPlot(seu_obj,group.by ="Phase" )
```

```{r}
p <- DimPlot(seu_obj,group.by ="Phase" )
p <- p + plot_annotation(
    title = "MLN Cell Cycle Effects Removed",
    theme = theme(
      plot.title = element_text(hjust = 0.5, size = 15, face = "bold")
    )
)

```

```{r}
save_pub(plot=p,  filename= "mln_cc_removed", width=7, height=6)
```


```{r}
rm(seu_obj)
```


```{r}
seu <- Seurat::RunPCA(seu, npcs = ifelse(ncol(seu) <= 50, ncol(seu) - 1, 50))
```



```{r}

saveRDS(
  object = seu,
  file = file.path(rds_dir, "seu_3k_mnn_pca_mln.rds")) # just in case
  
  
# Load it back
#seu <- readRDS(file.path(rds_dir, "seu_3k_mnn_pca_mln.rds"))
```


```{r}
#Taking from menglan
seu <- FindNeighbors(seu, reduction = "pca", dims = 1:30)
seu <- RunUMAP(seu, reduction = "pca", dims = 1:30, reduction.name = "umap.mnn")

```



```{r}

saveRDS(
  object = seu,
  file = file.path(rds_dir, "seu_umap_ran_mln.rds")) # just in case
  
  
# Load it back
#seu <- readRDS(file.path(rds_dir, "seu_umap_ran_mln.rds"))
```


```{r}
#Diagnostic Resolution
seu_obj <- Seurat::FindClusters(seu, resolution = seq(0.1, 0.8, by=0.1))

```

I can see from the output above that the resolution I need is 0.3 so gonna use that.



```{r}
seu_obj@graphs
```
```{r}
rm(seu_obj)
```

```{r}
seu <- FindClusters(seu, resolution = 0.3, cluster.name = "mnn_clusters")
```
```{r}

saveRDS(
  object = seu,
  file = file.path(rds_dir, "seu_mln_int_fin.rds")) # just in case
  
  
# Load it back
#seu <- readRDS(file.path(rds_dir, "seu_mln_int_fin.rds"))
```

```{r, fig.height=8, fig.width=12}

p <- DimPlot(seu, group.by="seurat_clusters", split.by = "condition",ncol=2, label =T) 
p <- p+ ggtitle("Viewing Cells in each cluster Split By Condition") +
  theme(plot.title = element_text(hjust = 0.5))
p
```
```{r}
save_pub(plot=p, filename="mln_splitby_cond", width= 12, height=8)
```


```{r}
p <- ggplot(seu@meta.data) +
    geom_bar(aes(x=orig.ident, fill=condition),
             stat="count", color="black") +
    theme_classic() +
    theme(axis.text.x = element_text(angle=45, vjust=1, hjust=1)) +
    labs(x="Sample name", y="Number of cells")
p
```

```{r}
save_pub(plot=p,filename= "mln_number_of_cells_viz")
```

{r Find best cluster resolution,fig.height=8,fig.width=10}

p <- clustree::clustree(seu_obj@meta.data[,grep("RNA_snn_res.", colnames(seu_obj@meta.data))],prefix = "RNA_snn_res.", layout="sugiyama")+ggtitle("Finding Best Resolution")+
  theme(
    plot.title = element_text(
      hjust = 0.5,       # centers the title
      margin = margin(b = 12)  # adds space (bottom margin)
    )
  )
p



save_pub(plot=p,  filename= "MLN_Cluster_determ", width=10, height=8)



r,fig.height=15,fig.width=7
Seurat::DimPlot(seu_obj, group.by = c("RNA_snn_res.0.1","RNA_snn_res.0.2", "RNA_snn_res.0.3"),label=TRUE, ncol=1)


Contaminant removal here


```{r}

#Contaminants
lec <- c("Prox1","Lyve1","Pdpn")
lymphocytes <- c("Ptprc","Cd52")
fib_ret <- c("Pdpn","Ccl19","Pdgfra")

```


```{r}
modules <- list(
  LEC = lec,
  Lymphocyte = lymphocytes,
  FRC = fib_ret
)

```

```{r}
DefaultAssay(seu) <- "RNA"
seu <- AddModuleScore(seu, features = modules)

```

```{r}
#Rename
old <- c("Cluster1", "Cluster2", "Cluster3")
new <- c("LEC", "Lymphocyte", "FRC")

colnames(seu@meta.data)[match(old, colnames(seu@meta.data))] <- new

```

```{r}
head(seu@meta.data)
```



```{r,fig.height=7,fig.width=8}
FeaturePlot(seu, features = c("LEC", "Lymphocyte", "FRC"),label=T,
            #cols = c("lightgrey", "red"),  # optional: nice contrast
            reduction = "umap.mnn")

```

Okay so we will do module score removal and then I will redo this contaminant removal analysis

```{r}
table(seu$orig.ident)
```
```{r,fig.width=12,fig.height=4}
p <- VlnPlot(
  seu,
  features = c("LEC", "Lymphocyte", "FRC"),
  group.by = "seurat_clusters",
  pt.size = 0
)
p <- p + plot_annotation(
    title = "MLN Contaminant Removal Analysis",
    theme = theme(
      plot.title = element_text(hjust = 0.5, size = 15, face = "bold")
    )
)

p
```

```{r}
save_pub(plot=p,  filename= "mln_Contaminant_removal_process", width=12, height=4)
```

So we can see that cluster 2 and 3 are lec and lec/frc respectively and 8 is lymphocyte but I will be a little conservative in my module score removal. We will obviously subset and remove.

```{r}
modules <- c("LEC", "Lymphocyte", "FRC")
thresholds <- c(0.6, 0.1, 0.1)   # example

plot_list <- list()

for (i in seq_along(modules)) {
    feat <- modules[i]
    thr <- thresholds[i]
    
    df <- data.frame(score = seu@meta.data[[feat]])
    
    p <- ggplot(df, aes(x = score)) +
        geom_histogram(bins = 60, fill = "lightgrey", color = "black") +
        geom_vline(xintercept = thr, color = "red", size = 1.2) +
        ggtitle(paste0("Histogram of MLN ", feat, " Score (Thr = ", thr, ")")) +
        theme_bw() +
        theme(plot.title = element_text(hjust = 0.5, size = 14, face = "bold"))
    
    plot_list[[i]] <- p
}


wrap_plots(plot_list, ncol = 1)

```


```{r}
p <- wrap_plots(plot_list, ncol = 1)
```


```{r}
save_pub(
  plot = p,
  filename = "module_histograms_mln"
)

```


```{r}
#keep cells that have a low score on these
seu <- subset(
  seu,
  subset = (LEC < 0.6) & (Lymphocyte < 0.1) & (FRC < 0.1))

```

```{r}
table(seu$orig.ident)
```
COUPOE-MLN             COUPWT-MLN MLN-COUPKO-HT3-cre-neg MLN-COUPKO-HT5-cre-pos 
                  7264                   5407                     81                    110 

```{r, fig.height=8, fig.width=12}

p <- DimPlot(seu, group.by="seurat_clusters", split.by = "condition",ncol=2, label =T) 
p <- p+ ggtitle("Viewing Cells in each cluster Split By Condition Post Module Removal") +
  theme(plot.title = element_text(hjust = 0.5))

```


```{r}
save_pub(plot=p, filename= "Cond_cluster_split_post_modules_subset",width=12,height=8)
```

```{r}

saveRDS(
  object = seu,
  file = file.path(rds_dir, "seu_mln_subset.rds")) # just in case
  
  
# Load it back
#seu <- readRDS(file.path(rds_dir, "seu_mln_subset.rds"))
```

```{r}
seu
```
An object of class Seurat 
22053 features across 12862 samples within 2 assays 
Active assay: RNA (19053 features, 3000 variable features)
 3 layers present: data, counts, scale.data
 1 other assay present: mnn.reconstructed
 3 dimensional reductions calculated: pca, integrated.mnn, umap.mnn

Now we integrate again!

```{r}
DefaultAssay(seu) <- "RNA"

seu <- NormalizeData(seu)
seu <- FindVariableFeatures(seu, nfeatures = 3000)

#cellcycle
s.genes = gorth(cc.genes.updated.2019$s.genes, source_organism = "hsapiens", target_organism = "mmusculus")$ortholog_name
g2m.genes = gorth(cc.genes.updated.2019$g2m.genes, source_organism = "hsapiens", target_organism = "mmusculus")$ortholog_name

seu <- CellCycleScoring(object=seu, s.features=s.genes, g2m.features=g2m.genes, set.ident=F) #set in the meta.data col phase
#Correcting by cc diff and viewing its effect
#This will rewrite the scale data
#seu$CC.Difference <- seu$S.Score - seu$G2M.Score

seu <- ScaleData(seu,assay="RNA", vars.to.regress = c("S.Score", "G2M.Score"))
#seu <- ScaleData(seu,assay="RNA", vars.to.regress = "CC.Difference")

seu <- RunPCA(seu, npcs = ifelse(ncol(seu) <= 50, ncol(seu) - 1, 50))

```




```{r}

saveRDS(
  object = seu,
  file = file.path(rds_dir, "seu_mln_subset_after_scale.rds")) # just in case
  
  
# Load it back
#seu <- readRDS(file.path(rds_dir, "seu_mln_subset_after_scale.rds"))
```

okay again i do wanna see if I have cell cycle removal? ah well expected anyway


```{r}

seu_obj <- RunPCA(seu, features = c(s.genes, g2m.genes))
DimPlot(seu_obj,group.by ="Phase" )
```

```{r}
rm(seu_obj)
```


```{r}
seu[["RNA"]] <- split(
  seu[["RNA"]],
  f = seu$orig.ident
)

```


```{r}
#batch=seu$orig.ident,
seu <- IntegrateLayers(seu, method = FastMNNIntegration,
                       features=VariableFeatures(seu),
                       new.reduction = "integrated.mnn",
                       verbose = TRUE)
```
```{r}
seu <- JoinLayers(seu)
```

```{r}
seu <- FindNeighbors(seu, reduction = "integrated.mnn", dims = 1:30)
seu <- RunUMAP(seu, reduction = "integrated.mnn", dims = 1:30, reduction.name = "umap.mnn")
```


```{r}

saveRDS(
  object = seu,
  file = file.path(rds_dir, "seu_mln_subset_after_umap.rds")) # just in case
  
  
# Load it back
#seu <- readRDS(file.path(rds_dir, "seu_mln_subset_after_umap.rds"))
```



```{r}

```






```{r}

```






```{r}

```

