---
title: "PP_2KO_part1"
author: "Ananya Gupta"
date: "2025-11-13"
output: html_document
---


```{r env_loading, results='FALSE',}
# Load packages, data and functions
library(Seurat)
library(tidyverse)
library(SeuratWrappers)
library(magicBatch)
library(BPCells)
library(batchelor)
library(patchwork)
library(Azimuth)
library(plotly)
library(gprofiler2)
library(hdf5r)
library(kableExtra)
library(ggplot2)
library(gridExtra)
library(DT)
library(clustree)
library(Matrix)
library(scales)
library(data.table)
library(dplyr)
library(tidyverse)
library(glue)
library(ggrepel)
library(gprofiler2)
library(multtest)
library(metap)

#for parallelization
options(future.globals.maxSize = 3000 * 1024^2) 
#options(future.globals.maxSize = 1e9) was abt 1GB


# Set the random number seed
set.seed(1234)
```

```{r dir setup, include=FALSE}
base_dir <- "PP_outputs"

# Define output directories
rds_dir  <- file.path(base_dir, "PP_all_rep_rds")
fig_dir  <- file.path(base_dir, "PP_all_rep_figs")

# Create folders if they don't exist
dir.create(base_dir, showWarnings = FALSE)
dir.create(rds_dir, showWarnings = FALSE)
dir.create(fig_dir, showWarnings = FALSE)
```
 saving a fig ggsave(
  filename = file.path(fig_dir, "cluster_umap.png"),
  plot = p,
  width = 6,
  height = 5,
  dpi = 300
)


saveRDS(
  object = seu,
  file = file.path(rds_dir, "seu_pp_all_reps.rds")
)

```{r}
library(Cairo)
save_pub <- function(plot,
                     filename,
                     width = 7,
                     height = 7,
                     dpi = 600) {

  

  # SAVE PDF
  ggsave(
    filename = file.path(fig_dir, paste0(filename, ".pdf")),
    plot = plot,
    device = CairoPDF,
    width = width,
    height = height,
    units = "in"
  )

  # SAVE PNG 
  CairoPNG(
    filename = file.path(fig_dir, paste0(filename, ".png")),
    width  = width * dpi,   # convert inches → pixels
    height = height * dpi,
    res    = dpi
  )

  print(plot) 
  dev.off()

  return(plot)
}

```


```{r}
samples_df <- read.csv("samples_pp_only.csv",header=TRUE, stringsAsFactors = FALSE) #gsub works on chr

samples_df$SampleName <- gsub("_", "-", samples_df$SampleName)

#fix path for windows
samples_df$Path <- trimws(samples_df$Path) #white space
samples_df$Path <-  normalizePath(samples_df$Path, winslash="/")#,mustWork = FALSE)


#setnames(values,namestoset)
datadirs <- setNames(samples_df[,4], samples_df[,1])


print(samples_df)
print(datadirs)
```

I will now create individual seurat object which I expect to see for the conditions/genotype =OE,WT,KO

In which we have 1 wt, 1ko and 3oe and 1 cre ctrl

We will integrate them based on sample?Or genotype?
Parse bio dataset was sample but I think u can do sondn too. Always do sample! https://www.biostars.org/p/9577962/

```{r}
seurat_list <- list()

for(i in 1:nrow(samples_df)) {
  sample_name <- samples_df$SampleName[i] #file names
  sample_group <- samples_df$SampleGroup[i] #genotype
  data_path <- samples_df$Path[i]
  
  cat("Processing:", sample_name, "\n")
  
  data <- Read10X(data.dir = data_path)
  
  #HTO data is multi output so it will be a list
  if(is.list(data)) {
    # For Cell Ranger multi output, extract only Gene Expression
    gene_exp <- data$`Gene Expression`
    cat("  - Multi output detected, using Gene Expression only\n")
  } else {
    # Regular output
    gene_exp <- data
    cat("  - Regular output\n")
  }
  
  # Create Seurat object
  seurat_obj <- CreateSeuratObject(
    counts = gene_exp,
    project = sample_name,
    min.cells = 3,
    min.features = 100
  )
  
  #metadata
  #seurat_obj$SampleName <- sample_name
  seurat_obj$genotype <- sample_group
  seurat_obj$site <- samples_df$Tissue[i]
  seurat_obj$condition <- samples_df$Condition[i]
  
  #qc
  seurat_obj$percent.mt <- PercentageFeatureSet(seurat_obj, pattern = "^mt-")

  
  # Store in list
  seurat_list[[sample_name]] <- seurat_obj
}

```

6 samples, 1KO,1WT, 2OE
```{r}
# look if it looks fine
head(seurat_list[[5]]@meta.data)
```


```{r}
# Merge all objects into one Seurat object
seu <- merge(
  x = seurat_list[[1]],
  y = seurat_list[2:length(seurat_list)],
  add.cell.ids = names(seurat_list),
  project = "PP_All_Reps",merge.data = TRUE
)
```

```{r}
print(Layers(seu))
seu
```

```{r}
saveRDS(
  object = seu,
  file = file.path(rds_dir, "seu_pp_just_merged.rds")
)
```

```{r}
seu <- readRDS(file.path(rds_dir, "seu_pp_just_merged.rds"))
```


```{r}
#Features above 6000 might be doublets ngl no but hevs have v high transcription
seu$genotype[seu$genotype == "KO_WT"] <- "CRE_CTRL"

p <- FeatureScatter(seu, feature1 = "nCount_RNA", feature2 = "nFeature_RNA", group.by="genotype") 
save_pub(plot=p, filename="mergeUMIvsgenes" )
  #+ ggtitle("QC: UMIs vs Genes")
```
```{r}
rm(seurat_list)
rm(seurat_obj)
```



```{r,fig.width=12}
df <- seu@meta.data
df$sum <- seu$nCount_RNA
df$detected <- seu$nFeature_RNA


p <- ggplot(df, aes(x = sum, y = detected)) +
  geom_point(aes(colour = percent.mt > 10), alpha = 0.7, size = 1) +
  facet_wrap(vars(genotype)) +
  scale_color_manual(
    values = c("FALSE" = "coral", "TRUE" = "turquoise3"),
    labels = c("FALSE" = "≤10%", "TRUE" = ">10%")
  ) +
  labs(
    title = "QC: Detected genes vs total counts per Peyer's Patches Sample",
    x = "Total counts per cell",
    y = "Detected genes per cell",
    colour = "Mitochondrial %"
  ) +
  theme_bw() +
  theme(
    plot.title = element_text(hjust = 0.5, face = "bold"),
    strip.text = element_text(face = "bold")
  )
save_pub(plot=p, filename="Detected_genes_vs_totcount",width=12)
```

```{r,fig.width=10, fig.height=4}
p <- VlnPlot(seu, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), group.by = "genotype",
              pt.size = 0,
              ncol = 3) #+ ggtitle("Before QC")
save_pub(plot=p, filename="b4_qc",width=10,height=4)
```

```{r}
#seu$orig.ident #sample name hai
```

```{r}
#cleaning
seu <- subset(seu, subset = nFeature_RNA > 100 & percent.mt < 10)
```


```{r}
saveRDS(
  object = seu,
  file = file.path(rds_dir, "seu_pp_merged_mt_clean.rds")
)
```

```{r}
seu <- readRDS(file = file.path(rds_dir,"seu_pp_merged_mt_clean.rds"))
```



```{r,fig.width=10, fig.height=4}
p <- VlnPlot(seu, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), group.by = "genotype",
              pt.size = 0,
              ncol = 3) #+ ggtitle("After QC")

# Visualize QC metrics as a violin plot
# VlnPlot(coupko_clean, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
save_pub(plot=p, filename="aft_qc",width=10,height=4)
```

Workflow to integrate

```{r}
seu <- NormalizeData(seu) #normalization.method = "LogNormalize",scale.factor = 10000
```

```{r}
#find 3000 var genes
seu <- FindVariableFeatures(seu, nfeatures = 3000)
                               #selection.method = "vst",
                               
```
Let us scale after int

as the results of this procedure are stored in the scaled data slot (therefore overwriting the output of ScaleData()) hence you can scale first look at plots and then regress out cell cycle effects

```{r}
seu <- ScaleData(seu) # Default is variable features that are getting scaled
```


https://satijalab.org/seurat/articles/cell_cycle_vignette
Regressed out pre integration ?
But https://github.com/satijalab/seurat/issues/1393 I think use it to look at it

can also set the identity of the Seurat object to the cell-cycle phase by passing set.ident = TRUE (the original identities are stored as old.ident).
restore by Idents(marrow) <- "old.ident" or any metadat col
```{r}
#Idents(seu) again sample name hai
```


This PCA is for int haha
```{r}

#marrow <- RunPCA(marrow, features = VariableFeatures(marrow), ndims.print = 6:10, nfeatures.print = 10)
#will be run using the variable features for the Assay. Note that the features must be present in the scaled data. Any requested features that are not scaled or have 0 variance will be dropped, and the PCA will be run using the remaining features.

#so run on scaled data

seu <- RunPCA(seu)
```

What if we just run it on the features?
Integration

```{r Integration}
#check idents
#Idents(seu)

# Check how many cells per batch
table(seu$orig.ident)

# Check if PCA has embeddings for all cells
ncol(seu)  # Total cells

```

```{r}
?FastMNNIntegration
```
https://github.com/satijalab/seurat/discussions/10083

FastMNN records its own pca dn not teh one thats being passed and hence is more resistant to batch eff and cell cycle 
```{r}
#seu <- IntegrateLayers(seu, method = FastMNNIntegration, batch=seu$orig.ident, features=3000, new.reduction = "integrated.mnn",verbose = TRUE)

seu <- IntegrateLayers(seu, method = FastMNNIntegration, batch=seu$orig.ident, new.reduction = "integrated.mnn",verbose = TRUE)


#
```

```{r}
head(seu[["integrated.mnn"]])
```

```{r}
seu <- JoinLayers(seu)
```

```{r}

saveRDS(
  object = seu,
  file = file.path(rds_dir, "seu_pp_int_joined_layers.rds")
)
# Load it back
#seu <- readRDS("seu_pp_joined_layers.rds")
```

```{r}
# Load it back
seu <- readRDS(file.path(rds_dir, "seu_pp_int_joined_layers.rds"))


```


note that if you wish to perform additional rounds of clustering after subsetting we recommend
re-running FindVariableFeatures() and ScaleData()

First lest do a diagnostic  resolution study


```{r}
seu <- FindNeighbors(seu, reduction = "integrated.mnn", dims = 1:30)
seu <- RunUMAP(seu, reduction = "integrated.mnn", dims = 1:30, reduction.name = "umap.mnn")
```
```{r}
#Diagnostic Resolution
seu_obj <- Seurat::FindClusters(seu, resolution = seq(0.1, 0.6, by=0.1))
```

```{r Find best cluster resolution,fig.height=8,fig.width=10}

clustree::clustree(seu_obj@meta.data[,grep("RNA_snn_res.", colnames(seu_obj@meta.data))],prefix = "RNA_snn_res.", layout="sugiyama")+ggtitle("Finding Best Resolution")+
  theme(
    plot.title = element_text(
      hjust = 0.5,       # centers the title
      margin = margin(b = 12)  # adds space (bottom margin)
    )
  )
```


```{r,fig.height=15,fig.width=7}
Seurat::DimPlot(seu_obj, group.by = c("RNA_snn_res.0.1","RNA_snn_res.0.2", "RNA_snn_res.0.3"),label=TRUE, ncol=1)
```
0.2.

```{r}
#remove the object you dont need
rm(seu_obj)
```

So now we do cell cycle, scale and then reductions

```{r}
# https://github.com/FrancisCrickInstitute/snRNAseq_workflow/blob/dabf3d87226ef9e9d51b12689c05a515cd1f0abc/templates/integrating_mnn.rmd
DefaultAssay(seu) <- "mnn.reconstructed"
```

```{r}
s.genes = gorth(cc.genes.updated.2019$s.genes, source_organism = "hsapiens", target_organism = "mmusculus")$ortholog_name
g2m.genes = gorth(cc.genes.updated.2019$g2m.genes, source_organism = "hsapiens", target_organism = "mmusculus")$ortholog_name
 
seu <- CellCycleScoring(object=seu, s.features=s.genes, g2m.features=g2m.genes, set.ident=F) #set in the meta.data col phase
```

```{r}
seu$CC.Difference <- seu$S.Score - seu$G2M.Score
seu <- ScaleData(seu, vars.to.regress = "CC.Difference")

seu <- RunPCA(seu, npcs = ifelse(ncol(seu) <= 50, ncol(seu) - 1, 50))
```

```{r}
ElbowPlot(seu, ndims=50)
```



https://github.com/icbi-lab/crc-atlas/blob/82c15ecc2ea36baa0c4cffe8818bf58e21dcfc48/analyses/50_Mouse_neutrophils_scRNAseq_analysis/3.1.Annotation_and_extraction_of_neutrophils.R


```{r}
saveRDS(
  object = seu,
  file = file.path(rds_dir, "seu_pp_int_cc_calc.rds")
)
```


```{r}
seu <- FindNeighbors(seu, reduction = "pca", dims = 1:30)
seu <- RunUMAP(seu, reduction = "pca", dims = 1:30, reduction.name = "umap.mnn")
seu <- FindClusters(seu, resolution = 0.2, cluster.name = "mnn_clusters")
```
```{r,fig.height=7,fig.width=7}
# View resuts
p2 <- DimPlot(seu, group.by="seurat_clusters") + NoLegend()
LabelClusters(p2, id = "seurat_clusters",  fontface = "bold", size = 5, bg.colour = "white", bg.r = .2, force = 0) + ggtitle("Integration with 3000 Features") + theme(plot.title = element_text(hjust = 0.5))
```

```{r,fig.height=12,fig.width=8}
DimPlot(seu, reduction = "umap.mnn", group.by = c("orig.ident", "condition", "genotype") ,ncol=1)
```

```{r,fig.width=10,fig.height=5}
DimPlot(seu, reduction = "umap.mnn", split.by = "condition")
```


```{r}

# Save your Seurat object
#saveRDS(seu, file = "seu_pp_int_fin.rds")


saveRDS(
  object = seu,
  file = file.path(rds_dir, "seu_pp_int_fin.rds")
)
# Load it back
#seu <- readRDS("seu_int.rds")
```







Older diagnostics


Lookin at hev diagnostics

```{r}
#  clust 11 has only G2M iterestinggggg 0.3 res

DimPlot(seu, reduction = "umap.mnn", group.by = "Phase") + ggtitle("Viewing Cell cycle Effects on Clusters") +
  theme(plot.title = element_text(hjust = 0.5))
```
lowkey theres no cell cycle effect?? DEAD AFFF

```{r}
FeaturePlot(seu, features = c("S.Score", "G2M.Score"), reduction = "umap.mnn")
FeaturePlot(seu, features = c("Mki67", "Ube2c", "Kif2c"), reduction = "umap.mnn")

```
```{r,fig.height=10, fig.width=9}
hev_markers <- c("St6gal1", "Glycam1", "Chst4", "Ccl21a", "Madcam1")
FeaturePlot(seu, features=hev_markers, reduction= "umap.mnn")#+ ggtitle("Viewing HEC Markers") +
  #theme(plot.title = element_text(hjust = 0.5))
```







```{r}
#PCA on CC before anything kind of regression out
# Running a PCA on cell cycle genes reveals, unsurprisingly, that cells separate entirely by
seu <- RunPCA(seu, features = c(s.genes, g2m.genes), reduction.name="pca.cellcycle.before", reduction.key="CC_B_")

```

```{r}
#Viewing the pca where theres no cell cycle correction
DimPlot(seu, reduction="pca.cellcycle.before",group.by = "Phase") +ggtitle("Before Cell cycle fixing") +
  theme(plot.title = element_text(hjust = 0.5))#before

```

```{r}
#now want to view what happens when you scale for the cell cycle scores in general

seu <- ScaleData(seu,assay="RNA", vars.to.regress = c("S.Score", "G2M.Score"))
seu <- RunPCA(seu, features = c(s.genes, g2m.genes), reduction.name="pca.cellcycle.score", reduction.key="CC_S_")

```

```{r}
DimPlot(seu, reduction="pca.cellcycle.score", group.by = "Phase") +ggtitle("Cell cycle fixing by Score") +
  theme(plot.title = element_text(hjust = 0.5))


```
Looks like cell cycle effects are not getting removed????? UMMMMMMMM maybe diff will help yk

```{r}
#Correcting by cc diff and viewing its effect
#This will rewrite the scale data
seu$CC.Difference <- seu$S.Score - seu$G2M.Score
seu <- ScaleData(seu,assay="RNA", vars.to.regress = "CC.Difference")
seu <- RunPCA(seu, features = c(s.genes, g2m.genes), reduction.name="pca.cellcycle.diff", reduction.key="CC_D_")

```

```{r}

#just to viz what happens after you correct for ccdiff
DimPlot(seu, reduction="pca.cellcycle.diff",group.by = "Phase") +ggtitle("Cell cycle fixing by Difference") +
  theme(plot.title = element_text(hjust = 0.5))

```
IT IS NOT GETTING REMOVED AFF PLS she just moving around im dead af
 Dgaf we just leave it in lowkey

Okay before any big moves lets do that Cell cycle viewing










!!! older stuff !!!


Run this only if you didnt run the stuff above

```{r}
seu <- FindNeighbors(seu, reduction = "integrated.mnn", dims = 1:35)

```

```{r}
seu <- FindClusters(seu, resolution = 0.3, cluster.name = "mnn_clusters")
```

```{r}
seu <- RunUMAP(seu, reduction = "integrated.mnn", dims = 1:35, reduction.name = "umap.mnn")
```


```{r}
head(seu@meta.data)
```

https://htmlpreview.github.io/?https://github.com/satijalab/seurat.wrappers/blob/master/docs/fast_mnn.html

```{r,fig.height=12,fig.width=8}
DimPlot(seu, reduction = "umap.mnn", group.by = c("orig.ident", "condition", "genotype"), ,ncol=1)
```

```{r,fig.width=10,fig.height=5}
DimPlot(seu, reduction = "umap.mnn", split.by = "condition")
```

```{r}
head(Idents(seu))
```


```{r}

# Save your Seurat object
#saveRDS(seu, file = "seu_pp_int.rds")


saveRDS(
  object = seu,
  file = file.path(rds_dir, "seu_pp_int.rds")
)
# Load it back
#seu <- readRDS("seu_int.rds")
```



```{r}

