---
title: "05_PP_DGE"
author: "Ananya Gupta"
date: "2025-11-19"
output: 05_HEV_PP_DGE
---

```{r env_loading, results='FALSE',}
# Load packages, data and functions
library(Seurat)
library(tidyverse)
library(SeuratWrappers)
library(magicBatch)
library(BPCells)
library(batchelor)
library(patchwork)
library(plotly)
library(gprofiler2)
library(hdf5r)
library(kableExtra)
library(ggplot2)
library(gridExtra)
library(DT)
library(clustree)
library(Matrix)
library(scales)
library(data.table)
library(dplyr)
library(tidyverse)
library(glue)
library(ggrepel)
library(gprofiler2)
library(multtest)
library(scater)
library(metap)
library(future)
library(parallelly)
library(scuttle)
library(SingleCellExperiment)
library(edgeR)
library(limma)
library(EnhancedVolcano)
library(fgsea)

library(AnnotationDbi)

library(org.Mm.eg.db)
library(clusterProfiler)
plan(multicore, workers = 8)


#for parallelization
options(future.globals.maxSize = 8000 * 1024^2) 
#options(future.globals.maxSize = 1e9) was abt 1GB


# Set the random number seed
set.seed(1234)
```


```{r dir setup, include=FALSE}
base_dir <- "PP_outputs"

# Define output directories
rds_dir  <- file.path(base_dir, "PP_all_rep_rds")
fig_dir  <- file.path(base_dir, "PP_all_rep_figs")
celltype_dir <- file.path(base_dir, "PP_HEV")
dir.create(celltype_dir, showWarnings = FALSE, recursive = TRUE)


# Create folders if they don't exist
dir.create(base_dir, showWarnings = FALSE)
dir.create(rds_dir, showWarnings = FALSE)
dir.create(fig_dir, showWarnings = FALSE)
```

For differential expression analysis of a multi-sample single-cell experiment, pseudo-bulk methods outperform DE methods speciĄcally designed at the single-cell level in terms of both
stability and computational speed [14].
We explore the pseudo-bulk proĄles using a multi-dimensional scaling (MDS) plot. This visualizes the differences between the expression proĄles of different samples in two dimensions.
It can be seen that pseudo-bulk samples from the same cell cluster are close to each other.


```{r}
seu <- readRDS(file.path(rds_dir, "seu_pp_7BEC_process_04res.rds"))
```

```{r}
DefaultAssay(seu)
```


```{r}
annotations <- read.csv("PP_outputs/top20_bec04res.csv")
```

```{r}

```
seu@meta.data$cluster_annot[seu@meta.data$seurat_clusters == 0] <- "CapEC2"
seu@meta.data$cluster_annot[seu@meta.data$seurat_clusters == 1] <- "Artery"
seu@meta.data$cluster_annot[seu@meta.data$seurat_clusters == 2] <- "Mafb+ Vn"
seu@meta.data$cluster_annot[seu@meta.data$seurat_clusters == 3] <- "TrEC"
seu@meta.data$cluster_annot[seu@meta.data$seurat_clusters == 4] <- "PCV"
seu@meta.data$cluster_annot[seu@meta.data$seurat_clusters == 5] <- "HEV"
seu@meta.data$cluster_annot[seu@meta.data$seurat_clusters == 6] <- "Pre-Artery"
seu@meta.data$cluster_annot[seu@meta.data$seurat_clusters == 7] <- "Vn Sele-"
seu@meta.data$cluster_annot[seu@meta.data$seurat_clusters == 8] <- "Barrier EC"
seu@meta.data$cluster_annot[seu@meta.data$seurat_clusters == 9] <- "CRP"

Before doing DGE, you should aggregate (combine) the expression values of cells from the same sample and the same cell type — using either sum, mean, or a mixed-model random effect. This aggregation step is called pseudobulk, and it is necessary because cells inside the same sample are correlated (not independent).

Lets start with HEV


```{r,fig.width=14,fig.height=5}
an2 <- DimPlot(seu, split.by= "condition",group.by="cluster_annot") + ggtitle("Annotated Peyers Patches") +
  theme(plot.title = element_text(hjust = 0.5))
LabelClusters(an2, id = "cluster_annot",  fontface = "bold", size = 2.5, bg.colour = "white", bg.r = .2, force = 0)
```


Important to create cell type pseudobulk

```{r}
pseudobulk_celltype <- function(
  seu, 
  celltype, 
  n_pseudo = 3,
  assay_name = "RNA",
  seed = 1234
) {
  # library(Seurat)
  # library(SingleCellExperiment)
  # library(dplyr)
  # library(scran)
  # library(scuttle)
  # library(scater)
  # library(limma)
  # 
  #----------------------
  # 1. Subset Seurat to that cell type
  #----------------------
  seu_sub <- subset(seu, subset = cluster_annot == celltype)
  message("Cells in this cell type: ", ncol(seu_sub))
  if (ncol(seu_sub) < 50) warning("Very few cells — PCA & DGE may be unstable.")
  
  #----------------------
  # 2. Convert to SCE
  #----------------------
  sce <- as.SingleCellExperiment(seu_sub, assay = assay_name)
  
  #----------------------
  # 3. Create pseudoreplicates per orig.ident
  #----------------------
  set.seed(seed)
  meta <- as.data.frame(colData(sce)) %>%
    mutate(cell = colnames(sce)) %>%
    group_by(orig.ident) %>%
    mutate(
      .rand = sample(seq_len(n())),
      pseudo_rep = rep(1:n_pseudo, length.out = n())[order(.rand)]
    ) %>%
    ungroup()
  
  colData(sce)$pseudo_rep <- meta$pseudo_rep[match(colnames(sce), meta$cell)]
  
  #----------------------
  # 4. Make pseudobulk IDs
  #----------------------
  colData(sce)$pb_id <- with(
    colData(sce),
    paste(orig.ident, paste0("rep", pseudo_rep), sep = "_")
  )
  
  #----------------------
  # 5. Aggregate to pseudobulk
  #----------------------
  pb <- aggregateAcrossCells(
    sce,
    ids = DataFrame(
      pb_id      = sce$pb_id,
      orig.ident = sce$orig.ident,
      condition  = sce$condition,
      pseudo_rep = sce$pseudo_rep
    )
  )
  
  # Save raw counts
  assay(pb, "counts_raw") <- assay(pb, "counts")
  pb$libsize <- colSums(assay(pb, "counts_raw"))
  
  #----------------------
  # 6. Normalize 
  #----------------------
  pb <- logNormCounts(pb)

  
  
  #----------------------
  # 8. Return everything
  #----------------------
  return(pb)
}

#use res_hev <- pseudobulk_celltype(seu, celltype = "HEV")

```

```{r}

#from pb make metadata,design matrix, dge list , filter, norm,disp and GLM
dge_prepare <- function(pb) {
  # library(edgeR)
  # library(dplyr)
  
  # Metadata extraction
  meta <- as.data.frame(colData(pb)) %>%
    dplyr::select(pb_id, condition, orig.ident)
  
  meta$condition <- factor(meta$condition)
  meta$orig.ident <- factor(meta$orig.ident)
  
  # Design matrix (no intercept)
  design <- model.matrix(~ 0 + condition , data = meta)
  
  # Build DGEList using raw pseudobulk counts
  y <- DGEList(
    counts = assay(pb, "counts_raw"),
    samples = meta
  )
  
  # Filter lowly expressed genes
  keep <- filterByExpr(y, design = design)
  y <- y[keep, , keep.lib.sizes = FALSE]
  
  # Normalize (TMM)
  y <- calcNormFactors(y)
  
  # Estimate dispersions
  y <- estimateDisp(y, design)
  
  # Fit GLM
  fit <- glmFit(y, design)
  
  # Return everything needed for contrasts
  return(list(
    fit = fit,
    design = design,
    y = y,
    meta = meta
  ))
}

```


##DGE
```{r}
pb_pcv <- pseudobulk_celltype(seu, celltype = "HEV")
hev_dge <- dge_prepare(pb_pcv)


```
```{r}
pcv_dge <- hev_dge
rm(hev_dge)
```


#pos fc means genes are higher in oe/ko/front neg means higher in wt 
```{r}
fit <- pcv_dge$fit
design <- pcv_dge$design

# 1. OE vs WT
c_oe_wt <- makeContrasts(conditionOE - conditionWT, levels = design)
res_oe_wt <- glmLRT(fit, contrast=c_oe_wt)
deg_oe_wt <- topTags(res_oe_wt, n = Inf)$table

# 2. KO vs WT
c_ko_wt <- makeContrasts(conditionKO - conditionWT, levels = design)
res_ko_wt <- glmLRT(fit,contrast= c_ko_wt)
deg_ko_wt <- topTags(res_ko_wt, n = Inf)$table

# 3. CRE_CTRL vs WT
c_cc_wt <- makeContrasts(conditionCRE_CTRL - conditionWT, levels = design)
res_cc_wt <- glmLRT(fit,contrast= c_cc_wt)
deg_cc_wt <- topTags(res_cc_wt, n = Inf)$table

# 4. OE vs KO
c_oe_ko <- makeContrasts(conditionOE - conditionKO, levels = design)
res_oe_ko <- glmLRT(fit,contrast= c_oe_ko)
deg_oe_ko <- topTags(res_oe_ko, n = Inf)$table
# 
# # 5. OE vs CRE_CTRL
# c_oe_cc <- makeContrasts(conditionOE - conditionCRE_CTRL, levels = design)
# res_oe_cc <- glmLRT(fit, contrast=c_oe_cc)
# deg_oe_cc <- topTags(res_oe_cc, n = Inf)$table
# 
# # 6. KO vs CRE_CTRL
# c_ko_cc <- makeContrasts(conditionKO - conditionCRE_CTRL, levels = design)
# res_ko_cc <- glmLRT(fit,contrast= c_ko_cc)
# deg_ko_cc <- topTags(res_ko_cc, n = Inf)$table


```

```{r}
deg_list <- list(
  OE_vs_WT  = deg_oe_wt,
  KO_vs_WT  = deg_ko_wt,
  CC_vs_WT  = deg_cc_wt,
  OE_vs_KO  = deg_oe_ko
#   OE_vs_CC  = deg_oe_cc,
#   KO_vs_CC  = deg_ko_cc
 )

```


OE WT


```{r,fig.width=9,fig.height=7}
EnhancedVolcano(
  deg_oe_wt,
  lab = rownames(deg_oe_wt),
  x = 'logFC',
  y = 'FDR',
  pCutoff = 0.05,
  FCcutoff = 1.2,     # log2 fold change cutoff of ±1
  pointSize = 2.0,
  labSize = 5,
  #col = c('grey30', 'grey30', 'royalblue', 'red2'),
  title = 'OE vs WT (HEV)',
  subtitle = 'DEG Volcano Plot',
  caption = 'Right = Upregulated | Left = Downregulated',
  legendPosition = 'right'
)
```

```{r}
# Significant DE genes
sig_deg <- deg_oe_wt %>% 
  filter(FDR < 0.05) 

# Up and down separately (optional)
up_deg <- sig_deg %>% filter(logFC > 0)
down_deg <- sig_deg %>% filter(logFC < 0)

# 
# library(AnnotationDbi)
# 
# library(org.Mm.eg.db)
# library(clusterProfiler)

gene_symbols <- rownames(sig_deg)

entrez_ids <- bitr(
  gene_symbols,
  fromType = "SYMBOL",
  toType = "ENTREZID",
  OrgDb = org.Mm.eg.db
)

sig_deg_entrez <- sig_deg %>% 
  mutate(SYMBOL = rownames(sig_deg)) %>%
  inner_join(entrez_ids, by = "SYMBOL")


ego_bp <- enrichGO(
  gene          = sig_deg_entrez$ENTREZID,
  OrgDb         = org.Mm.eg.db,
  keyType       = "ENTREZID",
  ont           = "BP",
  pAdjustMethod = "fdr",
  pvalueCutoff  = 0.05,
  readable      = TRUE
)

head(ego_bp)

```




ETS

```{r}
ranks <- deg_oe_wt$logFC
names(ranks) <- rownames(deg_oe_wt)
ranks <- sort(ranks, decreasing = TRUE)
library(msigdbr)

msig_c3 <- msigdbr(species = "Mus musculus", category = "C3", subcategory = "TFT")
msig_c2 <- msigdbr(species = "Mus musculus", category = "C2")


c3_tft <- split(msig_c3$gene_symbol, msig_c3$gs_name)
c2_pathways <- split(msig_c2$gene_symbol, msig_c2$gs_name)
#library(fgsea)

fgsea_ets <- fgsea(
  pathways = c3_tft,
  stats = ranks,
  minSize = 10,
  maxSize = 500
)

fgsea_ets <- fgsea_ets[grep("ETS|ERG|ELK|FLI|ETV|GGA", fgsea_ets$pathway), ]
fgsea_ets <- fgsea_ets[order(fgsea_ets$pval), ]
fgsea_ets


```


```{r}
#library(ggplot2)

top_ets <- fgsea_ets[order(fgsea_ets$padj), ] %>% 
  head(20)   # top 20 ETS pathways

ggplot(top_ets, aes(x = NES, y = reorder(pathway, NES))) +
  geom_point(aes(size = -log10(padj), color = NES)) +
  scale_color_gradient2(low = "blue", high = "red", mid = "white") +
  labs(
    title = "ETS-related enrichment (OE vs WT)",
    x = "Normalized Enrichment Score (NES)",
    y = "ETS / ERG / ELK / ETV target sets",
    size = "-log10(FDR)"
  ) +
  theme_bw()

```


KO WT

```{r,fig.width=9,fig.height=7}
EnhancedVolcano(
  deg_ko_wt,
  lab = rownames(deg_ko_wt),
  x = 'logFC',
  y = 'FDR',
  pCutoff = 0.05,
  FCcutoff = 1.2,     # log2 fold change cutoff of ±1
  pointSize = 2.0,
  labSize = 5,
  #col = c('grey30', 'grey30', 'royalblue', 'red2'),
  title = 'KO vs WT (HEV)',
  subtitle = 'DEG Volcano Plot',
  caption = 'Right = Upregulated | Left = Downregulated',
  legendPosition = 'right'
)
```

```{r}
#ibrary(dplyr)

# Top 20 OE-upregulated genes
top_up <- deg_ko_wt %>%
  filter(FDR < 0.05, logFC > 0) %>%
  arrange(desc(logFC)) %>% 
  head(20)

# Top 20 WT-upregulated genes (down in OE)
top_down <- deg_ko_wt %>%
  filter(FDR < 0.05, logFC < 0) %>%
  arrange(logFC) %>% 
  head(20)
write.csv(top_up,
          file.path(celltype_dir, "KO_vs_WT_top20_up.csv"),
          row.names = TRUE)

write.csv(top_down,
          file.path(celltype_dir, "KO_vs_WT_top20_down.csv"),
          row.names = TRUE)

```



```{r,fig.width=9,fig.height=7}

#goi <- c("Glycam1", "Ccl21a", "Selp", "Sele", "Vcam1", "Ackr1")   # example
goi <- c("Madcam1", "St6gal1") 



EnhancedVolcano(
  deg_ko_wt,
  lab = rownames(deg_ko_wt),
  x = 'logFC',
  y = 'FDR',
  pCutoff = 0.05,
  FCcutoff = 1.2,
  pointSize = 2.0,
  labSize = 4.0,
  selectLab = goi,                     # <-- highlight these genes
  #col = c('grey30','grey30','royalblue','red2'),
  title = 'KO vs WT (HEV)',
  subtitle = 'Madcam1, St6gal1',
  caption = 'Right = Upregulated | Left = Downregulated',
  legendPosition = 'right'
)


# EnhancedVolcano(
#   deg_oe_wt,
#   lab = rownames(deg_oe_wt),
#   x = 'logFC',
#   y = 'FDR',
#   pCutoff = 0.05,
#   FCcutoff = 1.2,     # log2 fold change cutoff of ±1
#   pointSize = 2.0,
#   labSize = 5,
#   #col = c('grey30', 'grey30', 'royalblue', 'red2'),
#   title = 'OE vs WT (PCV)',
#   subtitle = 'DEG Volcano Plot',
#   caption = 'Right = Upregulated | Left = Downregulated',
#   legendPosition = 'right'
# )
```
